Add program for deserializing from base64 code

piotr-now opened the pull request:
## Description
This PR conteins the new program 'ssl_base64_dump' dedicated to deserialize SSL context and session from the base64 code and print all obtained information on the stdout in a human-readable format. Base64 code should be delivered to the programs by a text file. The main operation of the program was based on the functions 'mbedtls_ssl_session_load ()' and 'mbedtls_ssl_session_load ()' from the code of the mbedTLS library. However, both functions could not be used due to the large dependencies with build configuration and context configuration when a serialization was created. This program was designed to work with different configuration of builds.


## Status
**READY** to review, but I would suggest to rebase all or some of the commits in to the one before merging. For now, I have kept the commit order for easier reviewing

## Requires Backporting
NO  

## Example of the output
```

Deserializing number 1:

Base64 code:
    AhMBAH8AAA8AAAGFAAAAAF5iHYPArgAgY98CXdeJ6TsLvGYSyFzVEmt5JDyMm0Tlt0bJjbrxZp3
    uZmmX4XcC9WOjcclhLqGKXg7FtgOk7cKvCcL1pSsrT+pHydC38U4fqTCG6YBJBnIAAAAAAAEYMII
    BFDCBuqADAgECAgReHs2iMAoGCCqGSM49BAMCMA0xCzAJBgNVBAMMAmNhMB4XDTIwMDExNTA4MzA
    yNloXDTQ1MDExNTA4MzAyNlowFzEVMBMGA1UEAwwMY29hcC1zZXJ2aWNlMFkwEwYHKoZIzj0CAQY
    IKoZIzj0DAQcDQgAEWVzJMmpJKZRTvmaWyRBCYr0EJ5R0OKB0t19fU4ananpdNH29WH3BRpirV8f
    8c1APq5ofUWYzWShwawH60anbazAKBggqhkjOPQQDAgNJADBGAiEAlLc6EJa+O9SWDGYv65G9a/T
    EuI6I7T1UXcyOcbSWKLUCIQD0kDm/2twx0f9TvihlZaVPk77zwSHDPGWP7BA1TCCkGQAAAAAAAAA
    AAABeYh2DG9FLjba+iGKkuBxpvozSY1+bLSzbnqCmLeJatl5iHYMDnC3cx9szLOfJPKt99xdr6cv
    zVlgDrErCNc2MAAgOrr3zFqxYiAAAAAAAAAAAAAAAAQAAAAAAAAADAAABAAAAAAACAAAA

Mbed TLS version:
    major    2
    minor    19
    path     1

Enabled sesion and context configuration:
    MBEDTLS_HAVE_TIME
    MBEDTLS_X509_CRT_PARSE_C
    MBEDTLS_SSL_MAX_FRAGMENT_LENGTH
    MBEDTLS_SSL_TRUNCATED_HMAC
    MBEDTLS_SSL_ENCRYPT_THEN_MAC
    MBEDTLS_SSL_SESSION_TICKETS
    MBEDTLS_SSL_SESSION_TICKETS and client
    MBEDTLS_SSL_DTLS_CONNECTION_ID
    MBEDTLS_SSL_DTLS_BADMAC_LIMIT
    MBEDTLS_SSL_DTLS_ANTI_REPLAY
    MBEDTLS_SSL_ALPN

Session info:
    start time     : 2020-03-06 09:53:07
    ciphersuite    : TLS-ECDHE-ECDSA-WITH-AES-128-CCM-8
    cipher flags   : 0x02
    cipher         : AES-128-CCM
    Message-Digest : SHA256
    compression    : disabled
    sesion ID      : 63 DF 02 5D D7 89 E9 3B 0B BC 66 12 C8 5C D5 12 
                     6B 79 24 3C 8C 9B 44 E5 B7 46 C9 8D BA F1 66 9D 
    master secret  : EE 66 69 97 E1 77 02 F5 63 A3 71 C9 61 2E A1 8A 
                     5E 0E C5 B6 03 A4 ED C2 AF 09 C2 F5 A5 2B 2B 4F 
                     EA 47 C9 D0 B7 F1 4E 1F A9 30 86 E9 80 49 06 72 
    verify result  : 0x00000000

Certificate:
    cert. version     : 3
    serial number     : 5E:1E:CD:A2
    issuer name       : CN=ca
    subject name      : CN=coap-service
    issued  on        : 2020-01-15 08:30:26
    expires on        : 2045-01-15 08:30:26
    signed using      : ECDSA with SHA256
    EC key size       : 256 bits

Session others:
    ticket lifetime          : 0 sec.
    MFL                      : none
    negotiate truncated HMAC : disabled
    Encrypt-then-MAC         : disabled

Random bytes:
    5E 62 1D 83 1B D1 4B 8D B6 BE 88 62 A4 B8 1C 69 BE 8C D2 63 5F 9B 
    2D 2C DB 9E A0 A6 2D E2 5A B6 5E 62 1D 83 03 9C 2D DC C7 DB 33 2C 
    E7 C9 3C AB 7D F7 17 6B E9 CB F3 56 58 03 AC 4A C2 35 CD 8C 

DTLS connection:
    in CID                             : none
    out CID                            : 0E AE BD F3 16 AC 58 88 
    bad MAC seen number                : 0
    last validated record sequence no. : 00 00 00 00 00 00 00 01 
    bitmask for replay detection       : 00 00 00 00 00 00 00 03 
    DTLS datagram packing              : disabled
    outgoing record sequence no.       : 00 01 00 00 00 00 00 02 
    MTU                                : 0
    ALPN negotiation                   : not selected

```

Patater commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -206,5 +234,7 @@ int main( int argc, char *argv[] )
         }
     }
 
+    printf( "Finish. Found %zu base64 codes\n", b64_counter );
```
I'm not sure we can use `%zu` on all our supported platforms. Might need to use inttypes.h printing macros, or something uglier.

piotr-now replied:
I changed all printed size_t values into uint32_t

Patater commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -205,22 +226,43 @@ size_t read_next_b64_code( char *b64, const size_t max_len )
+            ret = mbedtls_base64_decode( ssl, SSLBUF_LEN, &ssl_len, b64, b64_len );
+            if( ret != 0)
+            {
+                char buff[100];
```
Is 100 going to be enough? Consider using malloc to allocate the exact amount of bytes you need.

piotr-now replied:
This is a buffer only for error as a string and I do not think it wold be esy to detect how big should it be before calling `mbedtls_strerror()`. 100 should be good enough because that length is used in other parts of mbedTLS, but I will change it into 256 just in case.

piotr-now replied:
I removed `char buff[100];` here and used existing `b64_buf` instead

Patater commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -17,8 +17,8 @@
 
-static char debug = 0;
 FILE *b64_file = NULL;
+char debug = 0;
```
For commit "Fix typos", could you instead not add the typos into the commits where the typos originated?

Patater commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -292,6 +305,103 @@ void print_deserialized_ssl( /*const*/ unsigned char *ssl, size_t len )
+                                    context_cfg_flag );
+    ssl += session_len;
+
+     //TODO ssl_populate_transform
```
Style: indentation of this comment is off

Patater commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -115,9 +116,95 @@ void parse_arguments( int argc, char *argv[] )
+        if( len > 0)
+        {
+
+            //TODO: deserializing
```
Use C-style comments please

Patater commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -36,6 +38,10 @@
 char debug = 0;
+//TODO is keep_peer_certificate?
+char keep_peer_certificate = 1;
+//TODO is DTLS cmpiled?
```
```diff
- cmpiled
+ compiled
```

Patater commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -248,14 +254,55 @@ size_t read_next_b64_code( uint8_t *b64, size_t max_len )
 
+void print_deserialized_ssl_cert( const uint8_t *ssl, size_t len )
+{
+    enum { STRLEN = 4096 };
```
Is this going to be long enough? Could we allocate the exact right amount with malloc() instead?

piotr-now replied:
Unfortunately, I don't see a function that would check the required buffer size for a given certificate, but during my tests the output strings had around 300 characters, so I think 4096 is much more than required.

piotr-now replied:
the second thought is that `mbedtls_x509_crt_info` knows the size of the input buffer and it should be safe for memory

Patater commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -245,40 +247,51 @@ size_t read_next_b64_code( unsigned char *b64, size_t max_len )
+    int session_cfg_flag;
     int context_cfg_flag;
     size_t session_len;
+    //TODO is DTLS cmpiled?
```
```diff
- cmpiled
+ cmpiled
```

Patater commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -360,17 +359,54 @@ void print_deserialized_ssl_session( const uint8_t *ssl, size_t len,
+                    printf( "SHA512\n" );
+                    break;
+                case MBEDTLS_MD_RIPEMD160:
+                    printf( "RIPEMD160\n" );
```
Do we not have some cipher printing code in the library already that we could use?

piotr-now replied:
Yes, there is `mbedtls_md_info_from_type()`, but I decided to write my own due to the fact that `mbedtls_md_info_from_type()` will not work for algorithms with are turned off in the mbedTLS configuration

Patater commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -177,6 +182,14 @@ void print_hex( const uint8_t *b, size_t len )
 
+void print_time( const time_t *time )
+{
+    char buf[20];
```
Is 20 always enough?

piotr-now replied:
Yes, the output format has a constant length "2020-01-23 13:05:59"

Patater commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -177,6 +182,14 @@ void print_hex( const uint8_t *b, size_t len )
+{
+    char buf[20];
+    static const char * const format = "%G-%m-%d %H:%M:%S";
+    strftime( buf, 20, format, gmtime( time ) );
```
DRY (don't repeat yourself). Use sizeof(buf) to keep both places always in sync, and to show that the 20s have the same meaning.

Patater commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -185,6 +198,30 @@ void print_if_bit( const char *str, int bit, int val )
     }
 }
 
+const char * enable_disable( int val )
```
Style: Put `*` next to the identifier, here and in all other places in all other commits

```diff
- const char * enable_disable( int val )
+ const char *enable_disable( int val )
```

Patater commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -349,7 +430,7 @@ void print_deserialized_ssl_session( const uint8_t *ssl, size_t len,
                        ( (size_t) ssl[2] );
             ssl += 3;
-            printf_dbg( "cert_len: %zu\n", cert_len );
+            printf_dbg( "Certificate lenght: %zu\n", cert_len );
```
```diff
- lenght
+ length
```

Patater commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -33,11 +33,16 @@
 #include "mbedtls/ssl_ciphersuites.h"
 
-
+/*
```
Instead of commit "Add comments to the functions", please add comments to relevant code in the commits that add the code.

Patater commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,949 @@
+            printf( "\nBase64 code:\n" );
+            print_b64( b64, b64_len );
+
+            ret = mbedtls_base64_decode( ssl, SSLBUF_LEN, &ssl_len, b64, b64_len );
```
Before attempting to decode, I would recommend you check that the serialized data length is as expected and return an error if the base64 data is bigger than any valid data ever would be (in the biggest, fullest configuration), or smaller than any valid data ever would be (in the smallest configuration).

piotr-now replied:
I added checking length during reading base64 code from a file

Patater commented:
- Please add all.sh tests with example sessions to deserialize and grep for expected info from. Test with various library configurations and with valgrind and asan.
- Use all.sh to check that the program builds with CMake and Make, too.
- Please check that this builds and runs on Windows, too. You may need to create Visual Studio files, as this is a new program.

gilles-peskine-arm commented:
> Please add all.sh tests with example sessions to deserialize and grep for expected info from.

This goes beyond our current test coverage for sample programs, which is: if it builds, then ship it. But I'm glad we're finally doing this!

Unless you have a very good reason to do otherwise, please use the existing (pending review) framework, with `*_demo.sh` scripts: see https://github.com/ARMmbed/mbedtls/pull/2698

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+/*
+ *  SSL deserializer from base64 code
+ *
+ *  Copyright (C) 2006-2015, ARM Limited, All Rights Reserved
```
Please update this header with a good date.

piotr-now replied:
All programs in this folder have the same copyright date. It will change it of course, but please tell me what date should be placed here

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+#define TRANSFORM_RANDBYTE_LEN  64
+
+/*
+ * A macro that prevents from out of the ssl buffer range.
```
The comment is before an error string definition and lacks a verb.

piotr-now replied:
It was intended because that error string is a part of macro, but maybe it's better to move it to global values

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+#define CHECK_SSL_END( LEN )            \
+do                                      \
+{                                       \
+    if( end - ssl <  (int)( LEN ) )     \
```
Two spaces after `<`

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+char debug = 0;                         /* flag for debug messages */
+
+/*
+ * Base printing functions
```
```suggestion
 * Basic printing functions
```

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+    printf( "\nThis program is used to deserialize an mbedTLS SSL session from the base64 code provided\n"
+            "in the text file. The program can deserialize many codes from one file, but they must be\n"
+            "separated, e.g. by a newline.\n\n"
+            "Note: This program has been designed to work with mbedTLS v2.21.0 and may have some problems\n"
```
I have mixed feelings for this comment. Who should update this program if a breaking change is introduced? Will we have tests for this? Maybe this should be printed only if the detected version is other than 2.21? Maybe it should be dropped at all, since, if we have tests and everything works, there's nothing to worry about?

piotr-now replied:
I am not worry about future changes, but about the past, because we can always update the program and add backward compatibility. Creating tests for the current version of this program is a next step of work. This comment is more about old versions of mbedTLS than the current and future ones

The point here is that with this program you should be able to deserialize the ssl context from different library configurations and different versions.

AndrzejKurek replied:
Are you sure what backward compatibility you provide here? Is it tested? If not - we should not provide such warranty.

piotr-now replied:
I think full backward compatibility is unnecessary. Only for the current and future versions of mbedTLS. The guarantee will depend on well created tests

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+        "\t-h                 - Show this usage\n"
+        "\t-d                 - Print more information\n"
+        "\t--keep-peer-cert=0 - Use this option if you know that the mbedTLS library\n"
+        "\t                     has been compiled with the MBEDTLS_SSL_KEEP_PEER_CERTIFICATE\n"
```
Can't we test it by including the mbedTLS config?

piotr-now replied:
mbedTLS config during compilation will tell you nothing about serialized data because it may be made from a completely different configuration or even a different version of mbedTLS

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+        "\t-d                 - Print more information\n"
+        "\t--keep-peer-cert=0 - Use this option if you know that the mbedTLS library\n"
+        "\t                     has been compiled with the MBEDTLS_SSL_KEEP_PEER_CERTIFICATE\n"
+        "\t                     flag. You can also use it if there are some problems with reading\n"
```
What problems?

piotr-now replied:
mbedTLS serialize certificates in the diffrents way depends on MBEDTLS_SSL_KEEP_PEER_CERTIFICATE from configuration. This program should be able to desalinize both cases. By the default the deserialization proces works as if this flag was set, but if it wasn't, the deserialization will fail and errors will be displayed.
In that case the program users can set the '--keep-peer-cert=0' option to perform an alternative deserialization method.

This could be automated, but we would have to add information about this configuration flag to your serialized data i.e. make changes mbedTLS code

AndrzejKurek replied:
Please add a better description of what problems you expect. Writing "if there are some problems" in the documentation is very imprecise.

piotr-now replied:
The problem is that any error will depend on what data will be at the input. For example, when data is serialized without MBEDTLS_SSL_KEEP_PEER_CERTIFICATE (this program deserializes by default so that this configuration is set) - the program will read the length of the next fragment in the data when in fact there is completely different information. Then could be an error that there is not enough data or too much or other error during parsing subsequent data

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+        "\t                     has been compiled with the MBEDTLS_SSL_KEEP_PEER_CERTIFICATE\n"
+        "\t                     flag. You can also use it if there are some problems with reading\n"
+        "\t                     the information about certificate\n"
+        "\t--dtls-protocol=0  - Use this option if you know that the mbedTLS library\n"
```
Same as above - can't we include the config to know the defines?

piotr-now replied:
No, because we do not now if the serialized data was created with or without that config. The explanation is above

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+    {
+        va_list args;
+        va_start( args, str );
+        printf( "debug: ");
```
Inconsistent formatting of messages: `debug:` vs `ERROR:`

piotr-now replied:
This is intended. Error message should be more visible with a large amount of data at the output

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+ */
+void error_exit()
+{
+    if( b64_file )
```
If I'm not mistaken, there's a custom to be specific with NULL checks, so I would suggest 
```suggestion
    if( b64_file != NULL )
```
or, if you're fancy,
```suggestion
    if( NULL != b64_file )
```

piotr-now replied:
Actually, it's not only fancy, but it's also used as a good practice to reduce the probability of error :). For example, when you make a typo and don't write the double '=='. Then 'if (val = ENUM)' will cause runtime error and unforeseen behavior, while if (ENUM = val) will cause only compiling error

AndrzejKurek replied:
I do know that (that's why I said it's fancy), but considering the amount of reviews and the approach to code in the library, such error is highly unlikely, and the approach is not generally used (not sure if it's used anywhere in the library).

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+}
+
+/*
+ * This function print base64 code on the stdout
```
```suggestion
 * This function prints base64 code on the stdout
```

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+}
+
+/*
+ * This function print hex code from the buffer on the stdout.
```
```suggestion
 * This function prints hex code from the buffer on the stdout.
```

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+    printf("\t");
+    while( b < end )
+    {
+        if( ++i > 75 )
```
Why 75?

piotr-now replied:
Including "\t" it causes around 80 characters in one line

piotr-now replied:
I assumed that maximum in the line, but if we have one unified value one please let me know

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+/*
+ * Return pointer to hardcoded "enabled" or "disabled" depending on the input value
+ */
+const char * enable_disable( int val )
```
This function name is misleading - it doesn't enable or disable anything. I would rather suggest:
```suggestion
const char *print_enabled( int is_enabled )
```

AndrzejKurek replied:
Or, looking at how other functions are named, `get_enabled_str`

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+}
+
+/*
+ * Read next base64 code form the 'b64_file'. The 'b64_file' must be opened
```
```suggestion
 * Read next base64 code from the 'b64_file'. The 'b64_file' must be opened
```

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+ * place where it was previously end.
+ *
+ * /p b64       buffer for input data
+ * /p max_len   the maximum number of bytest to write
```
```suggestion
 * /p max_len   the maximum number of bytes to write
```

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+        {
+            if( missed > 0 )
+            {
+                printf_err( "Buffer for the base64 code is too low. Missed %zu characters\n", missed );
```
```suggestion
                printf_err( "Buffer for the base64 code is too small. Missed %zu characters\n", missed );
```

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+    {
+        char c_valid = 0;
+
+        c = (char)fgetc( b64_file );
```
Spaces missing

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+
+/*
+ * Read next base64 code form the 'b64_file'. The 'b64_file' must be opened
+ * previously. After next call of this function, the search will take from the
```
I don't really like the idea of a global file handle that's advanced with each read, and this function depending on it. In any way, this comment should rather state:

> After each call to this function, the internal file position indicator of the global `b64_file` is advanced.

piotr-now replied:
I do not think it is that advanced. It is just loading one character after another and interruption to parse current data packet

AndrzejKurek replied:
This comment was not addressed.

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+}
+
+/*
+ * This function deserialize and print at the stdout all obtained information
```
```suggestion
 * This function deserializes and prints to the stdout all obtained information
```

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+            if( 0 > ret )
+            {
+                mbedtls_strerror( ret, str, STRLEN );
+                printf_err( "Cannot write on the output - %s\n", str );
```
```suggestion
                printf_err( "Cannot write to the output - %s\n", str );
```

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+    {
+        mbedtls_strerror( ret, str, STRLEN );
+        printf_err( "Invalid format of X.509 - %s\n", str );
+        printf( "Cannot deserialize:\n\t" );
```
Isn't this an error?

piotr-now replied:
It was intended to use `printf` here instead of `printf_err` to avoid two "ERROR: " messages in two lines

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+}
+
+/*
+ * This function deserialize and print at the stdout all obtained information
```
```suggestion
 * This function deserializes and print to the stdout all obtained information
```

AndrzejKurek commented on programs/ssl/ssl_base64_dump.c:
```diff
@@ -0,0 +1,953 @@
+ * This function deserialize and print at the stdout all obtained information
+ * about the session from provided data. This function was built based on
+ * mbedtls_ssl_session_load(). mbedtls_ssl_session_load() could not be used
+ * due to dependencies on the mbedTLS configuration.
```
I would assume that this program will also be dependent on the mbedTLS configuration. Shouldn't it be?

piotr-now replied:
The assumption is that you have some data to deserialize and you don't know anything about it. You put it in to the program and you already know :)

piotr-now replied:
If this program depended too much too much on current mbedTLS configuration, you would be able to deserialize only data created by the same build

piotr-now commented:
> I still think that we should consider going one of two ways, either:
> 
> * Making this program config-agnostic, therefore not using `mbedtls_ssl_ciphersuite_from_id` (since it won't return a ciphersuite that's not compiled in) and other mbedtls functions that are affected by configuration. If we want to test this, the tests will fail anyway if their output is strictly tested and there are no dependencies defined. And if there will be dependencies defined, why not make it...
> * Tightly coupled to config. Only things that are in the configuration will be parsed. This is the way other programs work, and doing so makes the configuration easier.

The second option is the worst in my opinion because then we should have few builds of this program to cover all variants of configuration (almost 4 builds I can see now: MBEDTLS_SSL_KEEP_PEER_CERTIFICATE, MBEDTLS_SSL_PROTO_DTLS). Other programs use configuration to test current build of mbedTLS and this program should work with serialized data created by different builds. 

The best way in my opinion would be to make this program build with the max configuration enabled for cipher suites and not rewriting the contents of mbedtls_ssl_ciphersuite_from_id

piotr-now commented on programs/.gitignore:
```diff
@@ -45,6 +45,7 @@ ssl/ssl_mail_client
 ssl/ssl_pthread_server
 ssl/ssl_server
 ssl/ssl_server2
+ssl/ssl_base64_dump
```
Changed to `ssl_context_info`

AndrzejKurek commented on programs/ssl/ssl_context_info.c:
```diff
@@ -0,0 +1,1087 @@
+/*
+ *  SSL deserializer from base64 code
```
SSL context* deserializer

AndrzejKurek commented on programs/ssl/ssl_context_info.c:
```diff
@@ -0,0 +1,1087 @@
+/*
+ *  SSL deserializer from base64 code
+ *
+ *  Copyright (C) 2006-2015, ARM Limited, All Rights Reserved
```
2006-2020 :)

AndrzejKurek commented on programs/ssl/ssl_context_info.c:
```diff
@@ -0,0 +1,1087 @@
+#define MAX_BASE64_LEN      ( MAX_SERIALIZED_DATA * 4 / 3 + 3 )
+
+/*
+ * A macro that prevents from out of the ssl buffer range.
```
There is no verb here.

AndrzejKurek replied:
This comment has not been addressed.

AndrzejKurek commented on programs/ssl/ssl_context_info.c:
```diff
@@ -0,0 +1,1087 @@
+    printf( "\nThis program is used to deserialize an mbedTLS SSL session from the base64 code provided\n"
+            "in the text file. The program can deserialize many codes from one file, but they must be\n"
+            "separated, e.g. by a newline.\n\n"
+            "Note: This program has been designed to work with mbedTLS v2.21.0 and may have some problems\n"
```
I would remove this note once the tests are added.

piotr-now replied:
I would prefer to leave it for the fact that for older versions there will be no support and it can not be guaranteed that this program will work for everyone without any problems. However, I changed this note a bit

AndrzejKurek commented on programs/ssl/ssl_context_info.c:
```diff
@@ -0,0 +1,1087 @@
+ *        increases it to the maximum MAX_BASE64_LEN
+ *
+ * /p b64       pointer to the pointer of the buffer for input data
+ * /p max_len   pointer the maximum number of bytes to write. It can be changed
```
This is a very strange mechanism. Why pass a pointer to a buffer if we don't even know what will be written to it? Maybe this function should allocate a buffer itself and only set a pointer to a buffer that's passed as an argument?

piotr-now replied:
You're right here. However, I wanted to keep a reasonable responsibility for this buffer: in one place the buffer is created and removed, and in another place changed if necessary. It seems to me a better option than creating objects in one place and free it in another

piotr-now replied:
I changed the comment of this variable to "pointer to the current buffer capacity".

AndrzejKurek commented on programs/ssl/ssl_context_info.c:
```diff
@@ -0,0 +1,1087 @@
+ * /p max_len   pointer the maximum number of bytes to write. It can be changed
+ *              if the buffer needs to be increased
+ *
+ * \retval      number of bytes writen in to the b64 buffer or 0 in case no more
```
```suggestion
 * \retval      number of bytes written in to the b64 buffer or 0 in case no more
```

AndrzejKurek commented on tests/context-info.sh:
```diff
@@ -0,0 +1,400 @@
+#
+# Copyright (c) 2012-2020, ARM Limited, All Rights Reserved
+#
+# This program is intended for testing the ssl_cntext_info program
```
```suggestion
# This program is intended for testing the ssl_context_info program
```

AndrzejKurek commented on tests/context-info.sh:
```diff
@@ -0,0 +1,400 @@
+
+set -u
+
+# Go to directory with this script
```
I don't think that it's necessary to comment on a "cd" line :)

piotr-now replied:
Ok. removed

AndrzejKurek commented on programs/ssl/ssl_context_info.c:
```diff
@@ -0,0 +1,1094 @@
+            "in the text file. The program can deserialize many codes from one file, but they must be\n"
+            "separated, e.g. by a newline.\n\n"
+            "Note: This program has been designed to work with mbedTLS v2.21.0 and newer, so it may have\n"
+            "      some problems with deserializing context created by other mbedTLS versions.\n\n" );
```
Other versions, or older versions? You're mentioning that the program is designed to work with newer versions, so I would expect that it will only have problems with older versions.

AndrzejKurek commented on tests/context-info.sh:
```diff
@@ -0,0 +1,400 @@
+    echo ""                                                                   >> $2
+}
+
+# Print foot at the end of file <file path>
```
footer*

AndrzejKurek commented on tests/context-info.sh:
```diff
@@ -0,0 +1,400 @@
+    do
+        case $1 in
+            "-m")
+                if grep -v '^==' "$OUT_FILE" | grep -v 'Serious error when reading debug info' | grep -v "^*" | grep "$2" >/dev/null; then :; else
```
When and why do we ignore these errors?

piotr-now replied:
It is described above:
```
    # lines beginning with == are added by valgrind, ignore them
    # lines with 'Serious error when reading debug info', are valgrind issues as well
```

AndrzejKurek replied:
I understand these comments, but why do these errors happen and why do we ignore them? :)

piotr-now replied:
Because here we want do check only the output from ssl_context_info. Checking all logs as one could lead to misinterpretations. Logs from valgrind are checked separately and slightly earlier:
```
# check valgrind's results
    if [ "$USE_VALGRIND" -gt 0 ]; then
        if ! ( grep -F 'All heap blocks were freed -- no leaks are possible' "$OUT_FILE" &&
             grep -F 'ERROR SUMMARY: 0 errors from 0 contexts' "$OUT_FILE" ) > /dev/null
        then
            fail "Memory error detected"
            return
        fi
    fi
```

AndrzejKurek replied:
It would be great if this would be described in the line 150 - "ignore them, because we already checked them before"

piotr-now replied:
Added

AndrzejKurek commented on tests/context-info.sh:
```diff
@@ -0,0 +1,400 @@
+run_test "No MBEDTLS_SSL_KEEP_PEER_CERTIFICATE, server" \
+         "srv_no_keep_cert.txt" \
+         -arg "--keep-peer-cert=0" \
+         -m "Deserializing" \
```
Why only certain tests check for this string?

piotr-now replied:
I assumed that if deserialization will happen and without any error it is good enough to PASS, but I've added a few more conditions just in case

AndrzejKurek commented on tests/context-info.sh:
```diff
@@ -0,0 +1,400 @@
+         -u "The length of the base64 code found should be a multiple of 4" \
+         -n "bytes left to analyze from context"
+
+run_test "Too many data at the beginning of base64 code" \
```
too much*

AndrzejKurek commented on tests/context-info.sh:
```diff
@@ -0,0 +1,400 @@
+         -m "ERROR" \
+         -n "The length of the base64 code found should be a multiple of 4" \
+
+run_test "Too many data in the middle of base64 code" \
```
too much*

AndrzejKurek commented on tests/context-info.sh:
```diff
@@ -0,0 +1,400 @@
+         -m "ERROR" \
+         -n "The length of the base64 code found should be a multiple of 4" \
+
+run_test "Too many data at the end of base64 code" \
```
too much*

AndrzejKurek commented on programs/ssl/ssl_context_info.c:
```diff
@@ -0,0 +1,1094 @@
+
+    if( b64_counter > 0 )
+    {
+        printf_dbg( "Finish. Found %u base64 codes\n", b64_counter );
```
Finish -> Finished

AndrzejKurek commented on tests/data_files/base64/cli_no_alpn.txt:
```diff
@@ -0,0 +1,3 @@
+// MBEDTLS_SSL_ALPN
```
Please describe this in more detail.

AndrzejKurek commented on tests/data_files/base64/mtu_10000.txt:
```diff
@@ -0,0 +1,2 @@
+// mtu 100
```
is it 100 or 10000?

AndrzejKurek commented on tests/data_files/base64/srv_no_alpn.txt:
```diff
@@ -0,0 +1,2 @@
+// MBEDTLS_SSL_ALPN
```
Please describe this file in more detail.

AndrzejKurek commented on tests/scripts/all.sh:
```diff
@@ -1149,6 +1161,9 @@ component_test_asan_remove_peer_certificate () {
     if_build_succeeded tests/compat.sh
+
+    msg "test: context-info.sh, !MBEDTLS_SSL_KEEP_PEER_CERTIFICATE"
+    if_build_succeeded tests/context-inf.sh
```
This command will not even run. Please make sure that every test you add is manually run at least once.

piotr-now replied:
Why do you think? This program should work independently from MBEDTLS SSL_KEEP_PEER CERTIFICATE. That's why I added it here

AndrzejKurek replied:
```suggestion
    if_build_succeeded tests/context-info.sh
```

piotr-now replied:
Nice check :)

gilles-peskine-arm commented on programs/ssl/ssl_context_info.c:
```diff
@@ -0,0 +1,1094 @@
+    printf( "\nThis program is used to deserialize an mbedTLS SSL session from the base64 code provided\n"
+            "in the text file. The program can deserialize many codes from one file, but they must be\n"
+            "separated, e.g. by a newline.\n\n"
+            "Note: This program has been designed to work with mbedTLS v2.21.0 and newer, so it may have\n"
```
I don't see how we can promise that this version of the program will work with newer versions of Mbed TLS. They may have new features that this program doesn't recognize. Shouldn't this message say that the program is only designed to work with a matching version of Mbed TLS?

By the way, it's spelled “Mbed TLS”.

piotr-now replied:
You are the second person who pay attention to this, so I am eager to delete this note. Anyway, I mentioned it before. There were plans to add tests to create a serialized context from future versions of MbedTLS to detect future compatibility gaps

piotr-now commented on tests/context-info.sh:
```diff
@@ -0,0 +1,400 @@
+# This program is intended for testing the ssl_cntext_info program
+#
+
+set -u
```
No special reason. I just noticed that -e is not used in all scripts. I will add it

piotr-now commented on tests/context-info.sh:
```diff
@@ -0,0 +1,400 @@
+# Print test name <name>
+print_name() {
+    printf "$1 "
+    local LEN=$(( 72 - `echo "$1" | wc -c` ))
```
Czy to jakieś powszechn styl kodowania używany w MbedTLS? Pytam ponieważ ten fragment został zaczerpnięty z compat.sh

gilles-peskine-arm replied:
It's a common convention in sh: uppercase for environment variables, lowercase for non-exported shell variables. Or sometimes uppercase for global variables, lowercase for local variables (including variables intended to be local even if not declared local for portability). We haven't always followed it historically, however.

piotr-now commented on tests/context-info.sh:
```diff
@@ -0,0 +1,400 @@
+
+# Print test name <name>
+print_name() {
+    printf "$1 "
+    local LEN=$(( 72 - `echo "$1" | wc -c` ))
+    for i in `seq 1 $LEN`; do printf '.'; done
+    printf " "
```
I am not sure. It is very concise, but it seems less readable

piotr-now replied:
I will use it because it should work much faster.

piotr-now commented on tests/context-info.sh:
```diff
@@ -0,0 +1,400 @@
+
+# Print test name <name>
+print_name() {
+    printf "$1 "
```
Chyba, że ktoś użyje `%%`

piotr-now replied:
Unless someone uses `%%`

piotr-now commented on tests/data_files/base64/cli_cid.txt:
```diff
@@ -0,0 +1,2 @@
+// Client, CID 0xBEEF
```
Why to the Makefile? Not better for some README?
There was an idea to write a program that will generate a serialized context in the form of base64 code directly from the current MbedTLS version. The idea was to ensure that ssl_context_info would be compatible with the current version of MbedTLS in the future. I don't know if there are still such plans

gilles-peskine-arm replied:
There is a `tests/data_files/README`, but we found out the hard way that the instructions in that file tended not to be precise enough. Writing actual code that generates the data ensures that someone else can create an equivalent file later.

piotr-now replied:
These codes were generated by the ssl_clinet2 and ssl_serwer2 programs. With their help it is possible to perform connection with specific parameters and then do serialization of the context. I just added the existing `mbedtls_base64_encode()` function and displayed the created base64 code on the console. I do not think creating separate program is necessary.
Maybe should I add this function permanently to these programs?

Patater replied:
It wouldn't be too hard to add a new option to those programs to save their context out to a file with a format usable by the test. At least that much would be a big help in making more test cases later.

piotr-now commented on programs/ssl/ssl_context_info.c:
```diff
@@ -0,0 +1,1094 @@
+/*
+ * Minimum and maximum number of bytes for specific data: context, sessions,
+ * certificates, tickets and buffers in the program. The context and session
+ * size values have been calculated based on the 'print_deserialized_ssl_context()'
```
They have been calculated base on the comments and source codes from `print_deserialized_ssl_context()`, `print_deserialized_ssl_session()`, `mbedtls_ssl_context_save()` and `ssl_session_save()`. I did not see any other way how to check it. These magic values will end to be valid if the serialization / deserialization method changes without backward compatibility. If you know a better secure solution for the future I am open

piotr-now commented:
I think I have referred to all the comments. I see some CI fails, but they don't seem to make any changes. I see some CI errors, but they seem unrelated to my changes. On the other hand. not all of them are finish yet, so I will check them on a regular basis

piotr-now commented:
> @piotr-now - are you planning any additional changes, or is the PR ready for review? :)

Currently I am not planning any additional changes unless CI shows me some additional bugs to fix or I get some required changes. Please go ahead ;)

AndrzejKurek commented on programs/ssl/ssl_context_info.c:
```diff
@@ -0,0 +1,1109 @@
+                    printf( "RIPEMD160\n" );
+                    break;
+                default:
+                    printf( "undefined or error\n" );
```
```suggestion
                    printf( "undefined or erroneous\n" );
```
Error would suggest that this field describes an error :)

AndrzejKurek commented on tests/context-info.sh:
```diff
@@ -0,0 +1,431 @@
+    # check other assertions
+    # lines beginning with == are added by valgrind, ignore them
+    # lines with 'Serious error when reading debug info', are valgrind issues as well
+    # lines beginning with * are added by this scrip, ignore too
```
```suggestion
    # lines beginning with * are added by this script, ignore too
```

piotr-now commented:
Looks like CI failures are fixed now. There are still 4 failures but it looks like they are not related to my changes

AndrzejKurek commented:
Please comment on what changes have you added for this iteration - this will help with focusing the review.

piotr-now commented:
> Please comment on what changes have you added for this iteration - this will help with focusing the review.

Changes from last iteration are related with your requested changes and with CI failures: I added requiring configuration MBEDTLS_X509_CRT_PARSE_C and MBEDTLS_ERROR_C used by some of MbedTLS functions, added/updating Visual Studio 2010 project files, fixing incompatibilities between different compilers. The functionality of the program itself has not changed

 I believe that I have also referred to all previously suggested comments.

Patater commented on tests/context-info.sh:
```diff
@@ -0,0 +1,431 @@
+    printf "\nPASSED\n"
+else
+    printf "\nFAILED ( $T_FAILED / $T_COUNT tests )\n"
+fi
```
Small thing: could we always print the test count?

Patater commented:
CI failing Mbed OS only, which is a known issue and being worked on. Good to go.

gilles-peskine-arm commented on programs/ssl/ssl_context_info.c:
```diff
@@ -0,0 +1,1112 @@
+    int valid_balance = 0;  /* balance between valid and invalid characters */
+    size_t len = 0;
+    char pad = 0;
+    char c = 0;
```
This needs to be `int c`, since the variable needs to have room for every possible `char` value plus the special value `EOF`. As it stands the code stops reading the file on the first byte that matches the value of `EOF`. On most platforms that's -1 which doesn't match any valid base64 character, so it's only a minor bug (the file is silently truncated instead of ignoring the character), but there may be platforms where EOF has a different value that breaks well-formed files.

Any idea why the mingw build didn't complain about `EOF != c` which tries to compare an unsigned value (`char` is unsigned on Windows, right?) with the negative constant `EOF`? See https://github.com/ARMmbed/mbedtls/pull/3431/commits/69252ab39a39dfb6f1ba686ab92fa56e58b21e94
